tracing: introduce a minimal tracing mode and enable in tests

andreimatei opened the pull request:
This patch makes the Tracer's behavior more configurable. Before, the
Tracer would generally not create spans when there was no explicit
reason to do so, unless a testing knob option was passed to the Tracer
creation (TracerTestingKnobs{ForceRealSpans: true}). This patch elevates
this configuration to a proper tracer option. The Tracer can be
configured to always create spans (useful for having all the in-flight
operations represented in the active spans registry), to not do that
(the old behavior), or to read the configuration from an env var. This
env var is useful while I test various things with the goal of
eventually enabling this tracing mode everywhere. For now the env var
defaults to tracing being off, except in CrdbTestBuild's (i.e. unit
tests), where it's on.

Some tests are changed to explicitly request tracing to be off, so that
they retain the old behavior, as they were relying on it. Among them, a
logic tests needs it, so a new cluster configuration option is added to
logic tests and test clusters.

After this patch, I think we're pretty close to enabling this new
tracing mode in production. Before that though, there's a couple more
performance optimizations to make, and more testing against
span-use-after-Finish.

Release note: None

andreimatei committed:
jobs: remove picky test

This test checks that a job record is not updated with a trace id if
tracing the job is not enabled. This test fails if tracing is generally
enabled across the cluster. The code under test is this:
https://github.com/cockroachdb/cockroach/blob/96a125ade349979f7ff7a42e23f39aa544c5529f/pkg/jobs/adopt.go#L401-L408

And, I'm trying to make it such that tracing is always enabled across
the cluster. So, I'm deleting this test, and also morally accepting that
there will usually be an extra update to the job record. Hopefully,
always getting that trace id in the job is a good thing.

Release note: None

andreimatei committed:
tracing: introduce a minimal tracing mode and enable in tests

This patch makes the Tracer's behavior more configurable. Before, the
Tracer would generally not create spans when there was no explicit
reason to do so, unless a testing knob option was passed to the Tracer
creation (TracerTestingKnobs{ForceRealSpans: true}). This patch elevates
this configuration to a proper tracer option. The Tracer can be
configured to always create spans (useful for having all the in-flight
operations represented in the active spans registry), to not do that
(the old behavior), or to read the configuration from an env var. This
env var is useful while I test various things with the goal of
eventually enabling this tracing mode everywhere. For now the env var
defaults to tracing being off, except in CrdbTestBuild's (i.e. unit
tests), where it's on.

Some tests are changed to explicitly request tracing to be off, so that
they retain the old behavior, as they were relying on it. Among them, a
logic tests needs it, so a new cluster configuration option is added to
logic tests and test clusters.

After this patch, I think we're pretty close to enabling this new
tracing mode in production. Before that though, there's a couple more
performance optimizations to make, and more testing against
span-use-after-Finish.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73407)
<!-- Reviewable:end -->

andreimatei committed:
tracing: reduce use of WithForceRealSpan

Tests are using this option when creating spans when they care about
tracing. Now that the Tracer can easily be configured to always create
spans, that's generally a cleaner options for tests then repeating this
option with every span creation.

Release note: None

andreimatei commented:
Based on #73453 and #72993. Only the last 3 commits are to be reviewed here.
For commit `jobs: remove picky test`, I got a thumbs up from Aditya, so it can be ignored.

abarganier reviewed (approved):
This feels significant, in a good way, and I'm excited that we're heading in a direction where we think a tracing mode like this is possible from a performance perspective. I just had one question regarding a comment, but otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-:-MqLVBZE4jTbJicPgZXl:bsx5j5z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @andreimatei)

___
*[pkg/util/tracing/tracer.go, line 160 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-MqLUd094NkK1cqSrFNb:-MqLUd094NkK1cqSrFNc:b-clyain) ([raw file](https://github.com/cockroachdb/cockroach/blob/451f3503e5356089453d449a32c1df251ebe6994/pkg/util/tracing/tracer.go#L160)):*
> ```
> 	// ... If no
> 	// net.Trace/OpenTelemetry tracer is attached and a span is not explicitly
> 	// asked to record, spans do not record events.
> ```
This comment is slightly confusing to me. When we say "record events" here are we talking about log events (e.g. `Eventf` class of functions)? 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> I'm excited that we're heading in a direction where we think a tracing mode like this is possible from a performance perspective

You and me both. I think we're almost there.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-:-MqM6jZH2aJqzf-XHNM-:b-wulsla)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @abarganier)

___
*[pkg/util/tracing/tracer.go, line 160 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-MqLUd094NkK1cqSrFNb:-MqM5ah6DEOO4e9ha9Cm:b-hz5y9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/451f3503e5356089453d449a32c1df251ebe6994/pkg/util/tracing/tracer.go#L160)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

> ```
> 	// ... If no
> 	// net.Trace/OpenTelemetry tracer is attached and a span is not explicitly
> 	// asked to record, spans do not record events.
> ```
This comment is slightly confusing to me. When we say "record events" here are we talking about log events (e.g. `Eventf` class of functions)? 
</blockquote></details>

Right. "recording" refers to spans accumulating log messages (`log.Info/Eventf()` calls which turn into `span.Record()`), as well as accumulating a history of their finished children.
I've improved the comment some; please see now.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-:-MqNvWJy91DHRevVqQ7G:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @abarganier, and @andreimatei)

___
*[pkg/sql/logictest/logic.go, line 170 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-MqNtPvB8dhagvvDa55J:-MqNtPvB8dhagvvDa55K:bnjko2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/46a6d7229c55e8a6f20c1ecbf60f5a196b78abff/pkg/sql/logictest/logic.go#L170)):*
> ```Go
> //   enabled. This is equivalent to setting COCKROACH_EXPERIMENTAL_SPAN_CONFIGS.
> // - tracing-off: If specified, tracing defaults to being turned off. This is
> //   used to override the environment, which may ask for ttracing to be on by
> ```

ttracing


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73407)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @abarganier, and @andreimatei)

___
*[pkg/util/tracing/tracer.go, line 160 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-MqLUd094NkK1cqSrFNb:-MqPxOscEfkXYhX3sFX7:b-xat8sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/451f3503e5356089453d449a32c1df251ebe6994/pkg/util/tracing/tracer.go#L160)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Right. "recording" refers to spans accumulating log messages (`log.Info/Eventf()` calls which turn into `span.Record()`), as well as accumulating a history of their finished children.
I've improved the comment some; please see now.
</blockquote></details>

Thanks for clarifying, lgtm!


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73407)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15, @abarganier, and @andreimatei)

___
*[pkg/sql/logictest/logic.go, line 170 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-MqNtPvB8dhagvvDa55J:-MqUy4pD6gfQyqdrPY_y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/46a6d7229c55e8a6f20c1ecbf60f5a196b78abff/pkg/sql/logictest/logic.go#L170)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

ttracing
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73407#-:-MqVdFuM4ZUBCUgz4jA7:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15, @abarganier, and @andreimatei)



<!-- Sent from Reviewable.io -->
